src/librustc_typeck/astconv.rs

```diff
@@ -176,21 +187,369 @@ impl<'o, 'gcx: 'tcx, 'tcx> dyn AstConv<'gcx, 'tcx>+'o {
+                    (Some(&arg), Some(&param)) => {
+                        match (arg, &param.kind) {
+                            (GenericArg::Lifetime(_), GenericParamDefKind::Lifetime) => {
+                                push_kind(&mut substs, provided_kind(param, arg));
+                                params.next();
+                            }
+                            (GenericArg::Lifetime(_), GenericParamDefKind::Type { .. }) => {
+                                // We expected a type argument, but got a lifetime
+                                // argument. This is an error, but we need to handle it
+                                // gracefully so we can report sensible errors. In this
+                                // case, we're simply going to infer the remaining
+                                // arguments.
+                                args.by_ref().for_each(drop); // Exhaust the iterator.
+                            }
+                            (GenericArg::Type(_), GenericParamDefKind::Type { .. }) => {
+                                push_kind(&mut substs, provided_kind(param, arg));
+                                params.next();
+                            }
+                            (GenericArg::Type(_), GenericParamDefKind::Lifetime) => {
+                                // We expected a lifetime argument, but got a type
+                                // argument. That means we're inferring the lifetimes.
+                                push_kind(&mut substs, inferred_kind(None, param, infer_types));
+                                params.next();
+                                progress_arg = false;
+                            }
+                        }
+                    }
+                    (Some(_), None) => {
+                        // We should never be able to reach this point with well-formed input.
+                        // Getting to this point means the user supplied more arguments than
+                        // there are parameters.
+                    }
+                    (None, Some(&param)) => {
+                        // If there are fewer arguments than parameters, it means
+                        // we're inferring the remaining arguments.
+                        match param.kind {
+                            GenericParamDefKind::Lifetime => {
+                                push_kind(&mut substs, inferred_kind(None, param, infer_types));
+                            }
+                            GenericParamDefKind::Type { .. } => {
+                                let kind = inferred_kind(Some(&substs), param, infer_types);
+                                push_kind(&mut substs, kind);
+                            }
+                        }
+                        params.next();
+                    }
+                    (None, None) => break,
+                }
+                if progress_arg {
+                    args.next();
```

Comment:
Can you do `args.next()` manually in each arm? IMO it's easier to see how it progresses that way.